python/tank/util/metrics_cache.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (c) 2023 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+The Metrics logger thread is only launched once an engine is initialized.
+But we might want to log metrics event before Toolkit is fully initialized.
+
+For instance, when starting ShotGrid Desktop and authenticating, there is no
+engine set yet.
+
+This module provide a cache that can be filled prior to Toolkit initilization.
+Then, once the engine is set and the metric thread starts, the cache is consumed
+and send the items to the Metric instance.
+
+Unfortunatly, this cache can not live in a singleton in this module because,
+somehow, tk-core is fully unloaded off the python process between authentication
+and launch of an engine (desktopstartup).
+Instead, we store the records in the environment as JSON text
+"""
+
+import hashlib
+import json
+import os
+
+from . import metrics
+from .. import platform
+
+from .. import LogManager
+
+logger = LogManager.get_logger(__name__)
+
+
+def log(*args, **kwargs):
```

Comment:
Based on my previous comment, I think this function's signature can receive the two positional arguments instead to prevent any incompatibility with `EventMetric.log`.